pkg/gds/store/trtl/iterator.go

```diff
@@ -0,0 +1,281 @@
+	if i.err != nil {
+		return false
+	}
+
+	if i.index >= len(i.values)-1 && i.nextPageToken == "" {
+		// No more values from the iterator
+		return false
+	}
+
+	i.index++
+	if i.index > 0 && i.index < len(i.values) {
+		// We have already loaded the next value
+		return true
+	}
+
+	request := &trtlpb.IterRequest{
+		Namespace: i.namespace,
+	}
+	if i.index > 0 {
+		// We need to request the next page of values
+		request.Options = &trtlpb.Options{
+			PageToken: i.nextPageToken,
+		}
+	}
+
+	var reply *trtlpb.IterReply
+	var err error
+	ctx, cancel := withContext(context.Background())
+	defer cancel()
+
+	if reply, err = i.client.Iter(ctx, request); err != nil {
+		i.err = err
+		return false
+	}
+
+	// Add the new values to the slice
+	i.values = append(i.values, reply.Values...)
+	i.nextPageToken = reply.NextPageToken
+	return true
+}
+
+func (i *trtlBatchIterator) Prev() bool {
+	i.index--
+	if i.index < 0 {
+		i.index = 0
+		return false
+	}
+	return true
+}
+
+func (i *trtlBatchIterator) Seek(key []byte) bool {
+	// TODO: Figure out what to do about the batch seek method.
+	return false
+}
+
+func (i *trtlBatchIterator) Key() []byte {
+	return i.values[i.index].Key
+}
+
+func (i *trtlBatchIterator) Value() []byte {
+	return i.values[i.index].Value
+}
+
+func (i *trtlBatchIterator) Error() error {
+	return i.err
+}
+
+func (i *trtlBatchIterator) Release() {
+}
+
+// trtlStreamingIterator implements a streaming iterator for the trtl store.
+// The iterator fetches the next value from the server when the caller calls Next(),
+// and only the previous and current values are stored in memory. This is best suited
+// for use cases where snapshot isolation or one-at-a-time processing is required.
+type trtlStreamingIterator struct {
+	client    trtlpb.TrtlClient
+	cursor    trtlpb.Trtl_CursorClient
+	cancel    context.CancelFunc
+	prev      *trtlpb.KVPair
+	current   *trtlpb.KVPair
+	next      *trtlpb.KVPair
+	eof       bool
+	namespace string
+	err       error
+}
+
+func NewTrtlStreamingIterator(client trtlpb.TrtlClient, namespace string) *trtlStreamingIterator {
+	return &trtlStreamingIterator{
+		client:    client,
+		namespace: namespace,
+	}
+}
+
+func (i *trtlStreamingIterator) Next() bool {
+	if i.cursor == nil {
+		var ctx context.Context
+		ctx, i.cancel = withContext(context.Background())
+		request := &trtlpb.CursorRequest{
+			Namespace: i.namespace,
+		}
+		i.cursor, i.err = i.client.Cursor(ctx, request)
+	}
+
+	if i.err != nil {
+		return false
+	}
+
+	if i.next != nil {
+		// We have already loaded the next value
+		i.prev = i.current
+		i.current = i.next
+		i.next = nil
+		return true
+	}
+
+	if i.eof {
+		return false
+	}
+
+	// Fetch the next value from the cursor
+	var val *trtlpb.KVPair
+	if val, i.err = i.cursor.Recv(); i.err != nil {
+		if i.err == io.EOF {
+			i.err = nil
+			i.eof = true
+		}
+		return false
+	}
+
+	if i.current != nil {
+		i.prev = i.current
+	}
+
+	i.current = val
+
+	return true
```

Comment:
The `nil` check isn't necessary I don't think since we'd just be assigning `i.prev = nil`, right?

```suggestion
	// Bookkeeping to enable one `Prev` call 
	i.prev = i.current 
	i.current = val
	return true
```